File: helpers/monitor_sample.py
#!/usr/bin/env python3
"""Emit one `netdiag --monitor` sample as a single compact JSON line.

Called once per cycle from lib/monitor.sh with the sample's state exported
as NETDIAG_MON_* environment variables — the same pattern as
helpers/emit_json.py, and for the same reason: bash cannot escape a string
into JSON safely. An SSID may legally contain a quote, a backslash or a
newline, and a printf-built stream that the GUI parses forever would
eventually meet one. Interpreter startup costs ~50 ms once per cadence;
a malformed line costs the whole session.

The shape is deliberately *smaller* than a full `--json` run and is
documented separately in docs/JSON-SCHEMA.md. It is not a subset by
accident — a monitor sample answers "what is true right now", a run
answers "what is wrong and why".

Conventions match the full schema:
  * null means the probe did not run this cycle (its tier wasn't due, or
    there is no link). It never means "ran and measured zero".
  * `status.rules` are IDs from docs/DIAGNOSIS-RULES.md, evaluated in
    lib/monitor.sh against lib/thresholds.sh. Consumers render them; they
    do not re-derive them.
"""

from __future__ import annotations

import json
import os
import sys

# A broken sibling must not take the whole stream down: without the
# catalog we fall back to "Issue <id>" phrasing, which is worse prose
# but a live monitor.
try:
    from rules_catalog import RULES
except Exception:
    RULES = []

# Rule ID -> catalog title ("G2" -> "Router dropping packets"), so a
# rule-fired/rule-cleared summary speaks the same plain-English name the
# GUI's report card and `--rules-catalog` already use, rather than the bare
# rule ID a user has never seen. Built once at import time — RULES is a
# module-level constant, so this never changes within a process lifetime.
_RULE_TITLES: dict[str, str] = {r["id"]: r["title"] for r in RULES}


def _env(name: str) -> str | None:
    v = os.environ.get(f"NETDIAG_MON_{name}")
    if not v:
        return None
    # Foundation's JSONDecoder drops the whole line on a surrogate escape
    # (e.g. from a non-UTF-8 SSID) — a mojibake character beats a dropped
    # sample.
    return v.encode("utf-8", "replace").decode("utf-8")


def _f(name: str) -> float | None:
    v = _env(name)
    if v is None:
        return None
    try:
        return float(v)
    except ValueError:
        return None


def _i(name: str) -> int | None:
    v = _env(name)
    if v is None:
        return None
    try:
        return int(v)
    except ValueError:
        return None


def _tri(name: str) -> bool | None:
    """Three-valued: "1" true, "0" false, unset/empty None.

    _bool() would be wrong for every probe that can be skipped. Collapsing
    "not measured" to false is what turns a tier that hasn't run yet into
    "the internet is down" — the exact false-critical shape the packet-loss
    predicates in lib/common.sh exist to prevent.
    """
    v = os.environ.get(f"NETDIAG_MON_{name}", "")
    if v == "1":
        return True
    if v == "0":
        return False
    return None


def build_tcp() -> list[dict]:
    """NETDIAG_MON_TCP_LINES is one 'host|port|ok|elapsed_ms' per line."""
    raw = _env("TCP_LINES") or ""
    out: list[dict] = []
    for line in raw.splitlines():
        if not line.strip():
            continue
        parts = line.split("|")
        if len(parts) < 3:
            continue
        host, port, ok = parts[0], parts[1], parts[2]
        entry: dict = {
            "host": host,
            "port": int(port) if port.isdigit() else port,
            "ok": ok == "1",
            "elapsed_ms": None,
        }
        if len(parts) > 3 and parts[3]:
            try:
                entry["elapsed_ms"] = float(parts[3])
            except ValueError:
                pass
        out.append(entry)
    return out


def _changes() -> list[dict]:
    """Field-level diff against the previous sample (NETDIAG_MON_PREV_*).

    None on either side means "not measured" on that side, and an
    unmeasured→measured transition is not a change — same convention as
    the rest of the stream, where null is absence of measurement.
    Rules are the exception: they are always evaluated, so set
    difference is safe. Summaries are user-facing prose; the GUI
    renders them verbatim (CLAUDE.md: no verdict strings in Swift).

    Invariant this relies on: a field whose null suppresses the diff
    must keep its last known value in the previous-sample snapshot
    (bash side) — otherwise a single link-down sample with empty
    values erases the comparison baseline and interface/SSID changes
    are never reported.
    """
    if _env("HAVE_PREV") != "1":
        return []
    out: list[dict] = []

    def diff(now_key, prev_key, cid, field, phrase):
        now, prev = _env(now_key), _env(prev_key)
        if now is None or prev is None or now == prev:
            return
        out.append({"id": cid, "field": field, "from": prev, "to": now,
                    "summary": phrase(prev, now)})

    vpn_now = _env("VPN_ACTIVE") == "1"
    vpn_prev_raw = _env("PREV_VPN_ACTIVE")
    vpn_prev = vpn_prev_raw == "1"
    if vpn_prev_raw is not None and vpn_now != vpn_prev:
        name = _env("VPN_NAME") or _env("PREV_VPN_NAME")
        suffix = f" ({name})" if name else ""
        out.append({
            "id": "vpn-connected" if vpn_now else "vpn-disconnected",
            "field": "vpn.active",
            "from": "1" if vpn_prev else "0",
            "to": "1" if vpn_now else "0",
            "summary": (f"VPN connected{suffix}" if vpn_now
                        else f"VPN disconnected{suffix}"),
        })
    elif vpn_now and _env("VPN_TYPE") != "utun-route":
        # lib/monitor.sh sets MON_VPN_NAME to the tunnel interface
        # (utun4, utun6, ...) for utun-route VPNs — that's not a name,
        # and it duplicates interface-changed below.
        diff("VPN_NAME", "PREV_VPN_NAME", "vpn-name-changed", "vpn.name",
             lambda a, b: f"VPN changed: {a} → {b}")

    def exit_phrase(a, b):
        if vpn_now and vpn_prev:
            return f"VPN exit moved: {a} → {b}"
        if vpn_now and not vpn_prev:
            # Just connected: "a" was the user's real location, not a
            # prior VPN exit, so don't imply the exit itself moved.
            return f"VPN exit is in {b}"
        return f"Location changed: {a} → {b}"

    diff("PUB_CC", "PREV_PUB_CC", "country-changed", "public.country",
         exit_phrase)
    diff("PUB_IP", "PREV_PUB_IP", "public-ip-changed", "public.ip",
         lambda a, b: "Public IP changed")
    diff("PUB_ISP", "PREV_PUB_ISP", "isp-changed", "public.isp",
         lambda a, b: f"Internet provider changed: {a} → {b}")
    diff("SSID", "PREV_SSID", "wifi-network-changed", "link.ssid",
         lambda a, b: f"Wi-Fi network changed: {a} → {b}")
    if _env("SSID") is not None and _env("SSID") == _env("PREV_SSID"):
        diff("BSSID", "PREV_BSSID", "wifi-roamed", "link.bssid",
             lambda a, b: "Roamed to a different Wi-Fi access point")
    diff("INTERFACE", "PREV_INTERFACE", "interface-changed",
         "link.interface",
         lambda a, b: f"Network interface changed: {a} → {b}")

    rules_now = set((_env("RULES") or "").split())
    rules_prev = set((_env("PREV_RULES") or "").split())
    for rid in sorted(rules_now - rules_prev):
        out.append({"id": "rule-fired", "field": "status.rules",
                    "from": None, "to": rid,
                    "summary": _RULE_TITLES.get(rid, f"Issue {rid} detected")})
    for rid in sorted(rules_prev - rules_now):
        title = _RULE_TITLES.get(rid)
        out.append({"id": "rule-cleared", "field": "status.rules",
                    "from": rid, "to": None,
                    "summary": (f"Resolved: {title}" if title
                                else f"Issue {rid} cleared")})
    return out


def main() -> None:
    is_wifi = _env("IFACE_TYPE") == "wifi"
    link_up = os.environ.get("NETDIAG_MON_LINK_UP") == "1"
    rules = (_env("RULES") or "").split()

    sample = {
        # Fallback exists only for standalone/test invocation; it must
        # track NETDIAG_MON_SCHEMA in lib/monitor.sh.
        "schema": _i("SCHEMA") or 2,
        "version": _env("VERSION"),
        "ts": _env("TS"),
        "seq": _i("SEQ") or 0,
        # Which tiers actually refreshed this cycle. Everything outside
        # this list is carried over from an earlier sample, which a
        # consumer plotting a series needs to know before it draws a point.
        "refreshed": (_env("REFRESHED") or "").split(),
        "link": {
            "up": link_up,
            "interface": _env("INTERFACE"),
            "type": _env("IFACE_TYPE"),
            "ip": _env("LOCAL_IP"),
            "gateway": _env("GATEWAY"),
            "gateway_mac": _env("GW_MAC"),
            "ssid": _env("SSID"),
            "bssid": _env("BSSID"),
        },
        # Byte-identical to what a full run records, because lib/monitor.sh
        # calls lib/netid.sh rather than reimplementing precedence. This is
        # the join key between a live sample and the charted history.
        "network": {
            "id": _env("NETWORK_ID"),
            "label": _env("NETWORK_LABEL"),
        },
        "vpn": {
            "active": os.environ.get("NETDIAG_MON_VPN_ACTIVE") == "1",
            "type": _env("VPN_TYPE"),
            "name": _env("VPN_NAME"),
        },
        "gateway": {
            "loss_pct": _f("GW_LOSS"),
            "rtt_avg_ms": _f("GW_RTT"),
        },
        "internet": {
            "loss_pct": _f("INET_LOSS"),
            "rtt_avg_ms": _f("INET_RTT"),
        },
        "wifi": ({
            "rssi": _i("WIFI_RSSI"),
            "noise": _i("WIFI_NOISE"),
            "snr": _i("WIFI_SNR"),
            "channel": _env("WIFI_CHAN"),
        } if is_wifi else None),
        "dns": {
            "ok": _tri("DNS_OK"),
            "resolver": _env("DNS_RESOLVER"),
            "elapsed_ms": _f("DNS_MS"),
        },
        "tcp": {
            "any_ok": _tri("TCP_OK"),
            "targets": build_tcp(),
        },
        "public": {
            "ok": _tri("PUBLIC_OK"),
            "ip": _env("PUB_IP"),
            "isp": _env("PUB_ISP"),
            "asn": _env("PUB_ASN"),
            "city": _env("PUB_CITY"),
            "country": _env("PUB_CC"),
            "country_iso": _env("PUB_CC_ISO"),
            "captive_portal": _tri("CAPTIVE"),
        },
        "status": {
            "severity": _env("SEVERITY") or "ok",
            "rules": rules,
            # TCP-1 holding is the global suppressor for every loss alert.
            # Surfaced as its own boolean so a consumer doesn't have to
            # string-match the rules array to find it.
            "icmp_filtered": os.environ.get("NETDIAG_MON_ICMP_FILTERED") == "1",
            "degraded": os.environ.get("NETDIAG_MON_DEGRADED") == "1",
            # Paused by SIGUSR1 — probing is suspended, so every
            # measurement in this sample is stale by definition. A
            # consumer must not plot it or alert on it.
            "paused": os.environ.get("NETDIAG_MON_PAUSED") == "1",
            "cadence_s": _i("CADENCE_S"),
        },
    }

    changes = _changes()
    if changes:
        sample["changes"] = changes

    json.dump(sample, sys.stdout, separators=(",", ":"), default=str)
    sys.stdout.write("\n")
    # Flush per sample: this is a stream, and a consumer that has to wait
    # for a 4 KB pipe buffer to fill sees a status indicator that lags
    # reality by minutes.
    sys.stdout.flush()


if __name__ == "__main__":
    try:
        main()
    except BrokenPipeError:
        # The reader went away (the app quit, or `| head -5` closed the
        # pipe). Exit non-zero without a traceback so monitor_run's loop
        # sees it and stops cleanly instead of spinning against a dead fd.
        try:
            sys.stdout.close()
        except BrokenPipeError:
            pass
        sys.exit(1)

File: gui/Sources/NetdiagGUI/Models/MonitorSample.swift
import Foundation

/// One line of `netdiag --monitor`.
///
/// Every field is optional and every decode is lenient, for one reason: a
/// monitor sample is a *stream*, and a stream that throws on an unfamiliar
/// field stops the menu-bar indicator dead. The CLI may gain keys — it has
/// four times already — and the app must degrade to "I don't know that
/// yet" rather than to a blank icon.
///
/// Note what is *not* here: no thresholds, no severity computation, no
/// verdict strings. `status.rules` arrives pre-computed from
/// lib/monitor.sh, which reads the same lib/thresholds.sh that
/// lib/diagnosis.sh does. This type transports that decision; it never
/// makes one.
struct MonitorSample: Decodable, Sendable {
    var schema: Int?
    var version: String?
    var ts: String?
    var seq: Int?
    /// Which cadence tiers refreshed this cycle. Everything outside this
    /// list is carried over from an earlier sample — a chart drawing a
    /// point needs to know that before it plots one.
    var refreshed: [String] = []
    var link: Link = .init()
    var network: NetworkIdentity = .init()
    var vpn: VPN = .init()
    var gateway: Gateway = .init()
    var internet: Internet = .init()
    var wifi: WiFi?
    var dns: DNS = .init()
    var tcp: TCP = .init()
    var publicInfo: PublicInfo = .init()
    var status: Status = .init()

    /// Field-level differences from the previous sample, phrased by the
    /// CLI (schema 2+). Absent — and therefore empty — when nothing
    /// changed. `kind` is the stream's stable `id` string; `from`/`to`
    /// are nullable (rule transitions carry null on one side).
    var changes: [Change] = []

    struct Change: Decodable, Sendable, Equatable {
        var kind: String = ""
        var field: String?
        var from: String?
        var to: String?
        var summary: String = ""

        enum CodingKeys: String, CodingKey {
            case kind = "id"
            case field, from, to, summary
        }
    }

    enum CodingKeys: String, CodingKey {
        case schema, version, ts, seq, refreshed, link, network, vpn
        case gateway, internet, wifi, dns, tcp, status, changes
        case publicInfo = "public"
    }

    struct Link: Decodable, Sendable {
        var up: Bool = false
        var interface: String?
        var type: String?
        var ip: String?
        var gateway: String?
        var gatewayMAC: String?
        var ssid: String?
        var bssid: String?

        enum CodingKeys: String, CodingKey {
            case up, interface, type, ip, gateway, ssid, bssid
            case gatewayMAC = "gateway_mac"
        }

        var isWiFi: Bool { type == "wifi" }
    }

    struct NetworkIdentity: Decodable, Sendable, Equatable {
        var id: String?
        var label: String?

        enum CodingKeys: String, CodingKey {
            case id, label
        }
    }

    struct VPN: Decodable, Sendable {
        var active: Bool = false
        var type: String?
        var name: String?

        enum CodingKeys: String, CodingKey {
            case active, type, name
        }
    }

    struct Gateway: Decodable, Sendable {
        /// `nil` is "not measured this cycle", never zero. The whole
        /// null-vs-0 discipline in docs/JSON-SCHEMA.md exists because
        /// treating the first as the second produced false criticals.
        var lossPct: Double?
        var rttAvgMs: Double?

        enum CodingKeys: String, CodingKey {
            case lossPct = "loss_pct"
            case rttAvgMs = "rtt_avg_ms"
        }
    }

    struct Internet: Decodable, Sendable {
        var lossPct: Double?
        var rttAvgMs: Double?

        enum CodingKeys: String, CodingKey {
            case lossPct = "loss_pct"
            case rttAvgMs = "rtt_avg_ms"
        }
    }

    struct WiFi: Decodable, Sendable {
        var rssi: Int?
        var noise: Int?
        var snr: Int?
        var channel: String?

        enum CodingKeys: String, CodingKey {
            case rssi, noise, snr, channel
        }
    }

    struct DNS: Decodable, Sendable {
        /// Three-valued on purpose: `nil` means the medium tier has not
        /// run yet this session.
        var ok: Bool?
        var resolver: String?
        var elapsedMs: Double?

        enum CodingKeys: String, CodingKey {
            case ok, resolver
            case elapsedMs = "elapsed_ms"
        }
    }

    struct TCP: Decodable, Sendable {
        var anyOk: Bool?
        var targets: [Target] = []

        enum CodingKeys: String, CodingKey {
            case targets
            case anyOk = "any_ok"
        }

        struct Target: Decodable, Sendable {
            var host: String?
            var port: Int?
            var ok: Bool = false
            var elapsedMs: Double?

            enum CodingKeys: String, CodingKey {
                case host, port, ok
                case elapsedMs = "elapsed_ms"
            }
        }
    }

    struct PublicInfo: Decodable, Sendable, Equatable {
        var ok: Bool?
        var ip: String?
        var isp: String?
        var asn: String?
        var city: String?
        /// Full country name ("Brazil"). `countryISO` is the alpha-2 the
        /// flag rendering needs; both come from the CLI so nothing here
        /// has to ship a country table.
        var country: String?
        var countryISO: String?
        var captivePortal: Bool?

        enum CodingKeys: String, CodingKey {
            case ok, ip, isp, asn, city, country
            case countryISO = "country_iso"
            case captivePortal = "captive_portal"
        }
    }

    struct Status: Decodable, Sendable {
        var severity: String = "ok"
        /// Rule IDs from docs/DIAGNOSIS-RULES.md. The app maps these to
        /// alerts and renders them in the expert layer. It never decides
        /// whether one should have fired.
        var rules: [String] = []
        /// TCP-1 holds: real connections work, only ping is being dropped.
        /// The global suppressor for every loss alert — see AlertEngine.
        var icmpFiltered: Bool = false
        var degraded: Bool = false
        /// SIGUSR1 suspended probing. Every measurement in such a sample is
        /// carried over from before the pause, so a chart must not plot it
        /// and an alert must not fire on it.
        var paused: Bool = false
        var cadenceS: Int?

        enum CodingKeys: String, CodingKey {
            case severity, rules, degraded, paused
            case icmpFiltered = "icmp_filtered"
            case cadenceS = "cadence_s"
        }
    }

    var timestamp: Date { ISO8601DateFormatter().date(from: ts ?? "") ?? Date() }

    var health: Health {
        guard link.up else { return .critical }
        switch status.severity {
        case "critical": return .critical
        case "warn":     return .warning
        default:         return .healthy
        }
    }
}

/// Lenient, in an extension so the memberwise initializer survives — the
/// same discipline `RunSnapshot` uses, for a sharper reason. Swift's
/// synthesized decode throws `keyNotFound` rather than falling back to the
/// default written beside a property, so a netdiag one version behind on
/// any one of these keys would fail the decode of the *whole* sample and
/// stop the menu-bar indicator dead. `status.paused` is the current
/// example: it did not exist before the pause protocol.
extension MonitorSample {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        schema = c.lenient(.schema)
        version = c.lenient(.version)
        ts = c.lenient(.ts)
        seq = c.lenient(.seq)
        refreshed = c.lenient(.refreshed, [])
        link = c.lenient(.link, Link())
        network = c.lenient(.network, NetworkIdentity())
        vpn = c.lenient(.vpn, VPN())
        gateway = c.lenient(.gateway, Gateway())
        internet = c.lenient(.internet, Internet())
        wifi = c.lenient(.wifi)
        dns = c.lenient(.dns, DNS())
        tcp = c.lenient(.tcp, TCP())
        publicInfo = c.lenient(.publicInfo, PublicInfo())
        status = c.lenient(.status, Status())
        changes = c.lenient(.changes, [])
    }
}

extension MonitorSample.Change {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        kind = c.lenient(.kind, "")
        field = c.lenient(.field)
        from = c.lenient(.from)
        to = c.lenient(.to)
        summary = c.lenient(.summary, "")
    }
}

extension MonitorSample.Link {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        up = c.lenient(.up, false)
        interface = c.lenient(.interface)
        type = c.lenient(.type)
        ip = c.lenient(.ip)
        gateway = c.lenient(.gateway)
        gatewayMAC = c.lenient(.gatewayMAC)
        ssid = c.lenient(.ssid)
        bssid = c.lenient(.bssid)
    }
}

extension MonitorSample.NetworkIdentity {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = c.lenient(.id)
        label = c.lenient(.label)
    }
}

extension MonitorSample.VPN {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        active = c.lenient(.active, false)
        type = c.lenient(.type)
        name = c.lenient(.name)
    }
}

extension MonitorSample.Gateway {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        lossPct = c.lenient(.lossPct)
        rttAvgMs = c.lenient(.rttAvgMs)
    }
}

extension MonitorSample.Internet {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        lossPct = c.lenient(.lossPct)
        rttAvgMs = c.lenient(.rttAvgMs)
    }
}

extension MonitorSample.WiFi {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        rssi = c.lenient(.rssi)
        noise = c.lenient(.noise)
        snr = c.lenient(.snr)
        channel = c.lenient(.channel)
    }
}

extension MonitorSample.DNS {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        ok = c.lenient(.ok)
        resolver = c.lenient(.resolver)
        elapsedMs = c.lenient(.elapsedMs)
    }
}

extension MonitorSample.TCP.Target {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        host = c.lenient(.host)
        port = c.lenient(.port)
        ok = c.lenient(.ok, false)
        elapsedMs = c.lenient(.elapsedMs)
    }
}

extension MonitorSample.TCP {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        anyOk = c.lenient(.anyOk)
        targets = c.lenient(.targets, [])
    }
}

extension MonitorSample.PublicInfo {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        ok = c.lenient(.ok)
        ip = c.lenient(.ip)
        isp = c.lenient(.isp)
        asn = c.lenient(.asn)
        city = c.lenient(.city)
        country = c.lenient(.country)
        countryISO = c.lenient(.countryISO)
        captivePortal = c.lenient(.captivePortal)
    }
}

extension MonitorSample.Status {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        severity = c.lenient(.severity, "ok")
        rules = c.lenient(.rules, [])
        icmpFiltered = c.lenient(.icmpFiltered, false)
        degraded = c.lenient(.degraded, false)
        paused = c.lenient(.paused, false)
        cadenceS = c.lenient(.cadenceS)
    }
}

/// The menu-bar dot has three states. Deliberately not four: "info" is a
/// heads-up (a VPN is carrying your traffic), not a fault, and colouring it
/// differently from healthy would train the user to ignore the one colour
/// that matters.
enum Health: Sendable {
    case healthy, warning, critical

    var symbol: String {
        switch self {
        case .healthy:  return "circle.fill"
        case .warning:  return "exclamationmark.circle.fill"
        case .critical: return "xmark.circle.fill"
        }
    }

    /// What VoiceOver reads for the menu-bar item. The dot carries the whole
    /// state of the app in one glyph, so it needs words as well as a colour.
    var accessibilityLabel: String {
        switch self {
        case .healthy:  return "Network healthy"
        case .warning:  return "Network warning"
        case .critical: return "Network problem"
        }
    }
}
